java.lang.NullPointerException in AbstractCxxPublicApiVisitor.visitTemplateDeclaration

**Describe the bug**
SonarScanner fails to import source file because of  java.lang.NullPointerException. Analysis of same file through cxx-sslr-toolkit-2.0.2.2734.jar works without problems.

**To Reproduce**
Sadly I can't share this code. Also I'm not sure how to minimize it if sslr-toolkit doesn't crash. I've tried copying defines and includes but it doesn't change anything.

**Desktop (please complete the following information):**
 - OS: Windows
 - SonarQube version: 8.9.0
 - cxx plugin version: 2.0.2
 - sonar-scanner version:  tried on 4.6.2.2472 and sonar-scanner-cli-4.5.0.2216
 
**Additional context**
12:42:30.504 ERROR: Error during SonarScanner execution
org.sonar.cxx.squidbridge.api.AnalysisException: Unable to parse file: ....hxx
	at org.sonar.cxx.squidbridge.AstScanner.scanInputFiles(AstScanner.java:134)
	at org.sonar.plugins.cxx.CxxSquidSensor.execute(CxxSquidSensor.java:299)
	at org.sonar.scanner.sensor.AbstractSensorWrapper.analyse(AbstractSensorWrapper.java:48)
	at org.sonar.scanner.sensor.ProjectSensorsExecutor.execute(ProjectSensorsExecutor.java:49)
	at org.sonar.scanner.scan.ProjectScanContainer.doAfterStart(ProjectScanContainer.java:360)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:137)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:123)
	at org.sonar.scanner.bootstrap.GlobalContainer.doAfterStart(GlobalContainer.java:150)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:137)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:123)
	at org.sonar.batch.bootstrapper.Batch.doExecute(Batch.java:72)
	at org.sonar.batch.bootstrapper.Batch.execute(Batch.java:66)
	at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:46)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
	at com.sun.proxy.$Proxy0.execute(Unknown Source)
	at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:189)
	at org.sonarsource.scanner.api.EmbeddedScanner.execute(EmbeddedScanner.java:138)
	at org.sonarsource.scanner.cli.Main.execute(Main.java:112)
	at org.sonarsource.scanner.cli.Main.execute(Main.java:75)
	at org.sonarsource.scanner.cli.Main.main(Main.java:61)
Caused by: java.lang.NullPointerException
	at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitTemplateDeclaration(AbstractCxxPublicApiVisitor.java:655)
	at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitDeclarator(AbstractCxxPublicApiVisitor.java:527)
	at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitSingleDeclarator(AbstractCxxPublicApiVisitor.java:519)
	at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitDeclaratorList(AbstractCxxPublicApiVisitor.java:473)
	at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitNode(AbstractCxxPublicApiVisitor.java:410)
	at com.sonar.sslr.impl.ast.AstWalker.visitNode(AstWalker.java:114)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:85)
	at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
	at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
	at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
	at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
	at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
	at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
	at com.sonar.sslr.impl.ast.AstWalker.walkAndVisit(AstWalker.java:69)
	at org.sonar.cxx.squidbridge.AstScanner.walkAndVisit(AstScanner.java:155)
	at org.sonar.cxx.squidbridge.AstScanner.scanInputFiles(AstScanner.java:132)
	... 23 more

Hi @Nekto89,

thanks for your feedback.

The interesting part seems to be:
```
Caused by: java.lang.NullPointerException
at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitTemplateDeclaration(AbstractCxxPublicApiVisitor.java:655)
```
https://github.com/SonarOpenCommunity/sonar-cxx/blob/c59843eb36339248a85f2b254378857c961998c8/cxx-squid/src/main/java/org/sonar/cxx/visitors/AbstractCxxPublicApiVisitor.java#L655

```Java
    var declId = templateDeclaration.getFirstDescendant(CxxGrammarImpl.declaratorId);
    if (declId == null) {
      return;
    }
    var idNode = declId.getLastChild(qualifiedId);
    if (idNode != null) {
      idNode = idNode.getLastChild(IDENTIFIER);
    } else {
      idNode = declId;
    }
    String id = idNode.getTokenValue();
```

There seems to be a problem with a template declaration. Because the SSLR toolkit does not use the Public API visitor you don't get it there. Could you provide the template declaration creating the crash?

```
declaratorId
   LastChild => qualifiedId
      LastChild => IDENTIFIER => null
```

Regards,

--- a/cxx-squid/src/main/java/org/sonar/cxx/visitors/AbstractCxxPublicApiVisitor.java
+++ b/cxx-squid/src/main/java/org/sonar/cxx/visitors/AbstractCxxPublicApiVisitor.java
@@ -21,17 +21,16 @@ package org.sonar.cxx.visitors;
 
 import com.sonar.sslr.api.AstNode;
 import com.sonar.sslr.api.GenericTokenType;
-import static com.sonar.sslr.api.GenericTokenType.IDENTIFIER;
 import com.sonar.sslr.api.Grammar;
 import com.sonar.sslr.api.Token;
 import com.sonar.sslr.impl.ast.AstXmlPrinter;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.stream.Collectors;
 import javax.annotation.CheckForNull;
 import org.sonar.api.utils.log.Logger;
 import org.sonar.api.utils.log.Loggers;
 import org.sonar.cxx.parser.CxxGrammarImpl;
-import static org.sonar.cxx.parser.CxxGrammarImpl.qualifiedId;
 import org.sonar.cxx.parser.CxxKeyword;
 import org.sonar.cxx.parser.CxxPunctuator;
 import org.sonar.cxx.squidbridge.checks.SquidCheck;
@@ -646,13 +645,10 @@ public abstract class AbstractCxxPublicApiVisitor<G extends Grammar> extends Squ
     if (declId == null) {
       return;
     }
-    var idNode = declId.getLastChild(qualifiedId);
-    if (idNode != null) {
-      idNode = idNode.getLastChild(IDENTIFIER);
-    } else {
-      idNode = declId;
-    }
-    String id = idNode.getTokenValue();
+
+    String id = declId.getTokens().stream()
+      .map(Token::getValue)
+      .collect(Collectors.joining());
 
     // handle cascaded template declarations
     AstNode node = templateDeclaration;
@@ -815,7 +811,7 @@ public abstract class AbstractCxxPublicApiVisitor<G extends Grammar> extends Squ
       docNode = enumSpecifierNode;
     }
 
-    visitPublicApi(enumSpecifierNode, enumId,getBlockDocumentation(docNode));
+    visitPublicApi(enumSpecifierNode, enumId, getBlockDocumentation(docNode));
 
     // deal with enumeration values
     AstNode enumeratorList = enumSpecifierNode.getFirstDescendant(CxxGrammarImpl.enumeratorList);
@@ -841,7 +837,7 @@ public abstract class AbstractCxxPublicApiVisitor<G extends Grammar> extends Squ
           }
         }
 
-        visitPublicApi(definition,  definition.getFirstDescendant(  GenericTokenType.IDENTIFIER).getTokenValue(), comments);
+        visitPublicApi(definition, definition.getFirstDescendant(GenericTokenType.IDENTIFIER).getTokenValue(), comments);
       }
     }
   }

--- a/cxx-squid/src/test/java/org/sonar/cxx/visitors/CxxPublicApiVisitorTest.java
+++ b/cxx-squid/src/test/java/org/sonar/cxx/visitors/CxxPublicApiVisitorTest.java
@@ -83,7 +83,7 @@ public class CxxPublicApiVisitorTest {
 
   @Test
   public void template() throws IOException {
-    assertThat(verifyPublicApiOfFile("src/test/resources/metrics/template.h")).isEqualTo(tuple(14, 4));
+    assertThat(verifyPublicApiOfFile("src/test/resources/metrics/template.h")).isEqualTo(tuple(15, 4));
   }
 
   @Test

--- a/cxx-squid/src/test/resources/metrics/template.h
+++ b/cxx-squid/src/test/resources/metrics/template.h
@@ -75,3 +75,9 @@ struct A {
  * @brief issue #2138
  */
 template<> Formatter& LogMsg_applyFormat<int>(Formatter& format, int i);
+
+/**
+ * @brief issue #2180
+ */
+template <typename B>
+A<B>::~A() = default;

